subsystem/elevator.py

```diff
@@ -0,0 +1,204 @@
+            top_vertical_boundry_from_pivot = constants.vertical_boundary - constants.pivot_point_height - constants.top_boundry_buffer_gap
+            bottom_vertical_boundry_from_pivot = constants.pivot_point_height - constants.bottom_boundry_buffer_gap
+            horizontal_boundry_length_from_pivot = 0.5 * (constants.robot_length) + constants.horizontal_boundary - constants.side_boundry_buffer_gap
+            
+            top_boundry_max_angle = math.atan2(top_vertical_boundry_from_pivot, horizontal_boundry_length_from_pivot)
+            bottom_boundry_max_angle = math.atan2(bottom_vertical_boundry_from_pivot, horizontal_boundry_length_from_pivot) + math.radians(90)
+            # if the angle is within the minimun range of each of these boundry sides
+
+            # assuming the top is zero...
+            
+            if(angle < top_boundry_max_angle):
+                return get_length_ratio(angle, top_vertical_boundry_from_pivot)
+            elif(angle > top_boundry_max_angle and angle < bottom_boundry_max_angle):
+                return get_length_ratio(angle, horizontal_boundry_length_from_pivot)
+            elif (angle > bottom_boundry_max_angle):
+                return get_length_ratio(angle, bottom_vertical_boundry_from_pivot)
+
+    def set_length(self, length: meters):  # set arm extension
+        length_ratio = self.boundary_box(self.turn_sensor * (2 * math.pi))
+        if length_ratio * constants.max_elevator_height > length:
+            self.motor_extend.set_target_position(
+                length * constants.elevator_extend_gear_ratio)
+            self.elevator_length = self.motor_extend.get_sensor_position() / \
+                constants.elevator_extend_gear_ratio
+            return True
+        else:
+            self.motor_extend.set_target_position(
+                length_ratio * constants.max_elevator_height * constants.elevator_extend_gear_ratio)
+            return length_ratio * constants.max_elevator_height * constants.elevator_extend_gear_ratio
+
+    def shoulder_rotation_limits(self, radians: float):
+        # returns if the angle is within the soft limits
+        if self.rotation_overide:
+            return True
+        else:
+            if self.intake_up:
+                return radians < constants.shoulder_intake_up_max_rotation and radians > -constants.shoulder_min_rotation
+            else:
+                return radians < constants.shoulder_max_rotation and radians > -constants.shoulder_min_rotation
+
+    def set_rotation(self, radians: float):  # set arm rotation
+        # if the rotation is within the soft limits, set the rotation to the given angle
+        if self.soft_limits(radians):
+            self.main_rotation_motor.set_target_position(
+                radians * constants.elevator_rotation_gear_ratio)
+            # returns True
+            return True
+        else:
+            if self.intake_up:
```

Comment:
When you have a lot of similar lines right next to each other with only a few differences, that's a hint you can simplify your code. Since the only difference between those 7 lines in 1 line, by making that difference a variable, we can reuse those 7 lines twice instead of having them repeated.

```suggestion
            max_rotation = constants.shoulder_intake_up_max_rotation if self.intake_up else constants.shoulder_max_rotation
            if radians > 0:
                y = max_rotation
            else:
                y = -constants.shoulder_min_rotation
            self.main_rotation_motor.set_target_position(
            y * constants.elevator_rotation_gear_ratio)
            return y * constants.elevator_rotation_gear_ratio
```